apacheconfig/parser.py

```diff
@@ -89,44 +90,63 @@ def p_statement(self, p):
                 not hasattr(p[0][2], 'is_double_quoted')):
             p[0][2] = p[0][2].rstrip()
 
-    def p_statements(self, p):
-        """statements : statements statement
-                      | statement
+    def p_item(self, p):
+        """item : statement
+                | include
+                | includeoptional
+                | block
         """
-        n = len(p)
-        if n == 3:
-            p[0] = p[1] + [p[2]]
-        elif n == 2:
-            p[0] = ['statements', p[1]]
+        p[0] = p[1:][0]
+
+    def p_startitem(self, p):
+        """startitem : WHITESPACE item
```

Comment:
Just wondering, if it would be in the spirit of BNF to collapse some of the productions that we, handle in the same way, into sort/recurrent statements like these:

```
empty_space ::= WHITESPACE | NEWLINE | empty_space  # could we have a series of space/newline?
text ::= item | comment
contents : empty_space | text | contents
```
(this may not be a valid BNF for Ply, so it's just to explain what I am trying to say).

I am a bit worried by spreading comment/item/whitespace rules over three blocks of BNF statements. Would be possible to localize them somehow logically?

Reply:
Ah, to be a bit more clear (which I'll expand much more thoroughly in the comments)

In native Apache, parsing is done per-line, so even in-line comments are disallowed. For parsing only Apache configs, we'd require all items and comments to be split by `NEWLINE` only.

`Config::General`-type configs can have in-line comments, hence why comments are treated differently from `item`-- that is, `comment` can be preceded by any type of whitespace, whereas `item`s must be preceded by at least one newline (unless it is the first item in `contents`).

But there can be more simplification in the BNF (and many more comments!) to explain this more clearly.